src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -402,26 +402,57 @@ class GrowableArrayWithAllocator : public GrowableArrayView<E> {
 
   void push(const E& elem) { append(elem); }
 
-  E at_grow(int i, const E& fill = E()) {
+  template<typename FillConstructor>
+  E& at_grow_with(int i, FillConstructor fill_ctr) {
     assert(0 <= i, "negative index %d", i);
     if (i >= this->_len) {
-      if (i >= this->_capacity) grow(i);
-      for (int j = this->_len; j <= i; j++)
-        this->_data[j] = fill;
-      this->_len = i+1;
+      if (i >= this->_capacity) {
+        grow(i);
+      }
+
+      for (int j = this->_len; j <= i; j++) {
+        this->_data[j].~E();
+        fill_ctr(&this->_data[j]);
+      }
+
+      this->_len = i + 1;
     }
     return this->_data[i];
   }
 
-  void at_put_grow(int i, const E& elem, const E& fill = E()) {
+  template<typename... Args >
+  E& at_grow(int i, const Args&... args) {
+    return at_grow_with(i, [&](E* ptr) {
+      ::new (ptr) E(args...);
+    });
+  }
+
+  template<typename FillConstructor>
+  void at_put_grow_with(int i, const E& elem, FillConstructor fill_ctr) {
     assert(0 <= i, "negative index %d", i);
-    if (i >= this->_len) {
-      if (i >= this->_capacity) grow(i);
-      for (int j = this->_len; j < i; j++)
-        this->_data[j] = fill;
-      this->_len = i+1;
+    if (i < this->_len) {
+      this->_data[i] = E(elem);
```

Comment:
This is not good if we want to remove dependency on having an assignment operator. Maybe change the assignment and the return statement to a destructor call, and let the code path below initialize the value in the destructed memory?